File: src/hetzner_code/keys.py
"""Per-repo SSH keypairs. One keypair per (instance, repo) pair, never
reused across repos, so removing one repo's access can never affect
another's — see the design note in README.md.
"""

from __future__ import annotations

from pathlib import Path

from hetzner_code.run import run


def generate(key_dir: Path, *, comment: str) -> tuple[Path, Path]:
    """Generate a passwordless ed25519 keypair at key_dir/id_ed25519{,.pub}."""
    key_dir.mkdir(parents=True, exist_ok=True)
    private = key_dir / "id_ed25519"
    public = key_dir / "id_ed25519.pub"
    run(
        [
            "ssh-keygen",
            "-t",
            "ed25519",
            "-N",
            "",
            "-C",
            comment,
            "-f",
            str(private),
        ]
    )
    return private, public


def forget_private_key(private_key_path: Path) -> None:
    """Delete the local private key copy once it's safely on the box.
    The public key file is left in place — it's public, and handy for
    'which key did I register' debugging without another gh round trip."""
    if private_key_path.exists():
        private_key_path.unlink()

File: internal/keys/keys.go
// Package keys handles per-repo SSH keypairs. One keypair per
// (instance, repo) pair, never reused across repos, so removing one
// repo's access can never affect another's — see the design note in
// README.md.
package keys

import (
	"os"
	"path/filepath"

	"github.com/dikee/hetzner-code/internal/run"
)

// Generate creates a passwordless ed25519 keypair at
// keyDir/id_ed25519{,.pub}.
func Generate(keyDir, comment string) (privatePath, publicPath string, err error) {
	if err := os.MkdirAll(keyDir, 0o755); err != nil {
		return "", "", err
	}
	privatePath = filepath.Join(keyDir, "id_ed25519")
	publicPath = filepath.Join(keyDir, "id_ed25519.pub")
	_, err = run.Run(
		[]string{"ssh-keygen", "-t", "ed25519", "-N", "", "-C", comment, "-f", privatePath},
		"", true,
	)
	if err != nil {
		return "", "", err
	}
	return privatePath, publicPath, nil
}

// ForgetPrivateKey deletes the local private key copy once it's safely
// on the box. The public key file is left in place — it's public, and
// handy for "which key did I register" debugging without another gh
// round trip.
func ForgetPrivateKey(privateKeyPath string) error {
	err := os.Remove(privateKeyPath)
	if err != nil && os.IsNotExist(err) {
		return nil
	}
	return err
}
